rosbag2/src/rosbag2/types/ros2_message.cpp

```diff
@@ -0,0 +1,267 @@
+        auto data_vector = static_cast<std::vector<int32_t> *>(data);
+        std::vector<int32_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT32: {
+        auto data_vector = static_cast<std::vector<uint32_t> *>(data);
+        std::vector<uint32_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_INT64: {
+        auto data_vector = static_cast<std::vector<int64_t> *>(data);
+        std::vector<int64_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT64: {
+        auto data_vector = static_cast<std::vector<uint64_t> *>(data);
+        std::vector<uint64_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING: {
+        auto data_vector = static_cast<std::vector<std::string> *>(data);
+        std::vector<std::string> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE: {
+        auto data_vector = static_cast<std::vector<uint8_t> *>(data);
+
+        auto nested_ts =
+          static_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
+          member.members_->data);
+        auto size = member.size_function(data);
+
+        for (size_t j = 0; j < size; ++j) {
+          auto nested_member = data_vector->data() + j * nested_ts->size_of_;
+          deallocate_ros2_message_part(nested_member, nested_ts);
+        }
+
+        data_vector->resize(0);
+        data_vector->shrink_to_fit();
+        delete data_vector->data();
+        break;
+      }
+  }
+}
+
+void allocate_internal_types(
+  void * msg, const rosidl_typesupport_introspection_cpp::MessageMembers * members)
+{
+  for (size_t i = 0; i < members->member_count_; ++i) {
+    auto member = members->members_[i];
+    void * message_member = static_cast<uint8_t *>(msg) + member.offset_;
+
+    if ((member.is_array_ && member.array_size_ == 0) || member.is_upper_bound_) {
+      allocate_vector(message_member, member);
+    } else if (member.is_array_ && member.array_size_ > 0) {
+      allocate_array(message_member, member);
+    } else {
+      allocate_element(message_member, member);
+    }
+  }
+}
+
+void allocate_vector(void * data, rosidl_typesupport_introspection_cpp::MessageMember member)
+{
+  // This is necessary because initialization otherwise fails for MSVC++ compiled builds
+  if (member.type_id_ == rosidl_typesupport_introspection_cpp::ROS_TYPE_BOOL) {
+    new (data) std::vector<bool>();
+  }
+}
+
+void allocate_array(void * data, rosidl_typesupport_introspection_cpp::MessageMember member)
+{
+  if (member.type_id_ == rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING) {
+    auto string_array = static_cast<std::string *>(data);
```

Comment:
Maybe, however, we cannot cast to a `std::array<std::string, N> *` as we do not know `N` as a `const_expr`. So the cast to `std::string *` is teh best we can manage and it seems to work.

Reply:
> So the cast to std::string * is the best we can manage and it seems to work.

I'm not sure that's safe. Is it possible to instead store the `.data()` pointer (https://en.cppreference.com/w/cpp/container/array/data) along side the array? I don't think it's guaranteed that `std::array` starts with the data (as opposed to starting with other information that might be in the std::array class).

Seems like it is likely to work, but you're relying on implementation details and to me the behavior is undefined (therefore not safe): https://stackoverflow.com/a/39379359/671658

Reply:
I agree that our current solution is unsafe. 
The only alternative I can think of is to cast to `std::array<std::string, 1> *`, access only `data` and hope that N=1 is not special. This is probably a little bit more safe but IMHO also not guaranteed.
Would you prefer this @wjwwood?

I still think there is a underlying design problem in the cpp type support and instead of a pointer to the array a pointer to its data should be passed.

Reply:
That's still not safe because the layout of the array could depend on the value of `N`, for example `0` is probably a special case where the layout is different.

It's slightly safer from the perspective that I cannot think of a reason why `N=1` would have a different layout from `N=2` or more. But again, it's not guaranteed.

I think the right solution is to provide a pointer to `.data()` in the data structure or a function in the type support which can cast it to the right type and then return `.data()`.

Reply:
I completely agree. But changing the type support is not feasible for us right now.
However, we need a short term solution for rosbag.

Reply:
I would prefer to not support fixed sized arrays until it is fixed in the type support.

If you feel you must have this working in rosbag, then I guess you can use a work around, but honestly I think it's a bad idea.

Reply:
I found at least two places within ROS 2 where we actually do exactly that with our messages:

The first one is used in `fastrtps_dynamic` when parsing messages from CDR:
https://github.com/ros2/rmw_fastrtps/blob/master/rmw_fastrtps_dynamic_cpp/include/rmw_fastrtps_dynamic_cpp/TypeSupport_impl.hpp#L271-L287

We also do it in the message generation step ourselves:
https://github.com/ros2/rosidl/blob/master/rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.em#L43-L53

So I think this goes okay here as long as the message generation of `rosidl_typesupport_cpp` does the same thing.